src/PerpetualStableSwap.sol

```diff
@@ -0,0 +1,137 @@
+        emit ConditionalOrderCreated(_target);
+    }
+
+    function getTradeableOrder()
+        external
+        view
+        override
+        returns (GPv2Order.Data memory)
+    {
+        uint256 balanceA = tokenA.balanceOf(target);
+        uint256 balanceB = tokenB.balanceOf(target);
+
+        // Always sell whatever of the two tokens we have more of
+        IERC20 sellToken;
+        IERC20 buyToken;
+        uint256 sellAmount;
+        uint256 buyAmount;
+        if (convertAmount(tokenA, balanceA, tokenB) > balanceB) {
+            sellToken = tokenA;
+            buyToken = tokenB;
+            sellAmount = balanceA;
+            buyAmount = convertAmount(tokenA, balanceA, tokenB)
+                .mul(BPS.add(halfSpreadBps))
+                .div(BPS);
+        } else {
+            sellToken = tokenB;
+            buyToken = tokenA;
+            sellAmount = balanceB;
+            buyAmount = convertAmount(tokenB, balanceB, tokenA)
+                .mul(BPS.add(halfSpreadBps))
+                .div(BPS);
+        }
+        require(sellAmount > 0, "not funded");
+
+        // Unless spread is 0 (and there is no surplus), order collision is not an issue as sell and buy amounts should
+        // increase for each subsequent order. We therefore set validity to a large time span
+        // Note, that reducing currenbt block to a common start time is needed so that the order returned here
+        // does not change between the time it is queried and the time it is settled.
+        uint32 validity = 1 weeks;
+        // solhint-disable-next-line not-rely-on-time
+        uint32 currentTimeBucket = ((uint32(block.timestamp) / validity) + 1) *
+            validity;
+        return
+            GPv2Order.Data(
+                sellToken,
+                buyToken,
+                target,
+                sellAmount,
+                buyAmount,
+                currentTimeBucket + validity,
+                keccak256("PerpetualStableSwap"),
+                0,
+                GPv2Order.KIND_SELL,
+                false,
+                GPv2Order.BALANCE_ERC20,
+                GPv2Order.BALANCE_ERC20
+            );
+    }
+
+    function convertAmount(
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken
+    ) public view returns (uint256 destAmount) {
+        uint8 srcDecimals = srcToken.decimals();
+        uint8 destDecimals = destToken.decimals();
+
+        if (srcDecimals > destDecimals) {
+            destAmount = srcAmount.div(10 ** (srcDecimals.sub(destDecimals)));
+        } else {
+            destAmount = srcAmount.mul(10 ** (destDecimals.sub(srcDecimals)));
+        }
+    }
+
+    /// @param orderDigest The EIP-712 signing digest derived from the order
+    function isValidSignature(
+        bytes32 orderDigest,
+        bytes calldata
+    ) external view override returns (bytes4) {
+        require(
+            ConditionalOrder(this).getTradeableOrder().hash(domainSeparator) ==
```

Comment:
Shouldn't this use `target`? The downside of using `target` is that you also are forced to use a `CALL` instead of a more gas-efficient `JUMP` (internal function call).

Reply:
I was getting the following error when trying to call getTradeableOrder() directly:

> DeclarationError: Undeclared identifier. "getTradeableOrder" is not (or not yet) visible at this point.

Reply:
I think you need to `s/external/public` for that to allow both external and internal calls.